test/foundry/new/helpers/MOATHelpers.sol

```diff
@@ -0,0 +1,276 @@
+    }
+
+    /**
+     * @dev Get the "state" of the given order.
+     * @param order a MOATOrder.
+     * @param seaport a SeaportInterface, either reference or optimized.
+     */
+    function getState(
+        MOATOrder memory order,
+        SeaportInterface seaport
+    ) internal view returns (State) {
+        uint256 counter = seaport.getCounter(order.order.parameters.offerer);
+        bytes32 orderHash = seaport.getOrderHash(
+            order.order.parameters.toOrderComponents(counter)
+        );
+        (
+            bool isValidated,
+            bool isCancelled,
+            uint256 totalFilled,
+            uint256 totalSize
+        ) = seaport.getOrderStatus(orderHash);
+
+        if (totalFilled != 0 && totalSize != 0 && totalFilled == totalSize)
+            return State.FULLY_FILLED;
+        if (totalFilled != 0 && totalSize != 0) return State.PARTIALLY_FILLED;
+        if (isCancelled) return State.CANCELLED;
+        if (isValidated) return State.VALIDATED;
+        return State.UNUSED;
+    }
+
+    /**
+     * @dev Get the "type" of the given order.
+     * @param order a MOATOrder.
+     */
+    function getType(MOATOrder memory order) internal pure returns (Type) {
+        OrderType orderType = order.order.parameters.orderType;
+        if (
+            orderType == OrderType.FULL_OPEN ||
+            orderType == OrderType.PARTIAL_OPEN
+        ) {
+            return Type.OPEN;
+        } else if (
+            orderType == OrderType.FULL_RESTRICTED ||
+            orderType == OrderType.PARTIAL_RESTRICTED
+        ) {
+            return Type.RESTRICTED;
+        } else if (orderType == OrderType.CONTRACT) {
+            return Type.CONTRACT;
+        } else {
+            revert("MOATEngine: Type not found");
+        }
+    }
+
+    /**
+     * @dev Get the "structure" of the given order.
+     *
+     *      Note: Basic orders are not yet implemented here and are detected
+     *      as standard orders for now.
+     *
+     * @param order a MOATOrder.
+     */
+    function getStructure(
+        MOATOrder memory order
+    ) internal pure returns (Structure) {
+        // If the order has extraData, it's advanced
+        if (order.order.extraData.length > 0) return Structure.ADVANCED;
+
+        // If the order has numerator or denominator, it's advanced
+        if (order.order.numerator != 0 || order.order.denominator != 0) {
+            return Structure.ADVANCED;
+        }
+
+        (bool hasCriteria, bool hasNonzeroCriteria) = _checkCriteria(order);
+        bool isContractOrder = order.order.parameters.orderType ==
+            OrderType.CONTRACT;
+
+        // If any non-contract item has criteria, it's advanced,
+        if (hasCriteria) {
+            // Unless it's a contract order
+            if (isContractOrder) {
+                // And the contract order's critera are all zero
+                if (hasNonzeroCriteria) {
+                    return Structure.ADVANCED;
```

Comment:
STANDARD

Reply:
Can you take another look at this when you have a chance? I think the logic here might actually be correct, there's just a few layers of negatives that make it difficult to read.

From the doc:
- Order cannot contain any items with itemType ERC721WithCriteria or ERC1155WithCriteria, *unless*:
    - The order is a contract order and
    - All criteria-based items have `identifierOrCriteria == 0`
    - The contract offerer can successfully generate an order that resolves every criteria-based item without any context / extraData

```solidity
        // Order cannot contain any items with itemType ERC721WithCriteria or ERC1155WithCriteria, *unless*
        if (hasCriteria) {
            // The order is a contract order
            if (isContractOrder) {
                // and all criteria-based items have `identifierOrCriteria == 0`
                if (hasNonzeroCriteria) {
                    return Structure.ADVANCED;
                }
            } else {
                return Structure.ADVANCED;
            }
        }
```

Because these are basically guards, the default case here is `Structure.STANDARD`. If the logic is right, I'll try and detangle it—checking if there are nonzero criteria in an "unless" case is a confusing way to do this.

Reply:
ah that makes sense!